### ISSUE TITLE
Respawn event being called when the player is still dead

### ISSUE THREAD

**Issue by bloggy** [SkriptLang/Skript#287]
Please read my thread here:
https://forums.skunity.com/topic/7743/teleport-after-respawn-is-inconsistent/2

This bug has been around since forever I think. It's really annoying not being able to have skript execute all effects in their correct way and order when using the "on respawn" event.

It's not only the teleport stuff not working correctly in an "on respawn" event, but also many other methods. Sometimes adding a "wait x ticks" right after the "on respawn" event helps but not always (e.g. when there is a tps drop the moment the "on respawn" event is fired). 
There should not be any need for adding a "wait x ticks" at all in an "on respawn" event. 

I hope you can fix this :-)

**Comment by bensku** [comment]
You know, there are many operations which are not possible when player is dead. If you do not wait in respawn event, player has not respawned yet and you cannot for example teleport them. Even after waiting 1 tick there is no guarantee by Spigot that the player is alive...

I could technically make Skript inherently wait 1 tick, but as said, it might not even help anything.

**Comment by bloggy** [comment]
The "on respawn" event should only trigger if the player IS ALIVE. It makes no sense that it gets triggered when the player is still dead, or am I wrong? So every effect I am using in the "on respawn" event should be executed correctly without any issues.

**Comment by Pyon-Pyon** [comment]
@bloggy besku can't do anything for change that, it's only triggered when the event is called, so, this is more an issue of Spigot rather than Skript.

**Comment by bloggy** [comment]
It's just ... I've never seen any plugins that had such glitches when using the on respawn event

**Comment by Pyon-Pyon** [comment]
I didn't had any issues with that event, even if I use three or more of them in different triggers. And your issue is about a tps drop which could glitch any plugin no matter what.

**Comment by bensku** [comment]
Most plugins basically *do* use delay of some sort, or maybe simply do not try to perform unsafe operations like teleportation

**Comment by bloggy** [comment]
This bug still exists:
https://forums.skunity.com/threads/on-respawn-not-triggering.4978/

**Comment by JeepSmash** [comment]
I do not use the On Respawn event at all anymore.  I have switched to just using on death:
```
on death of a player:
  victim has permission "special.respawn.perm"
  loop while victim isn't alive:
    wait 1 tick
    victim isn't online
    stop
  <insert code to execute upon respawn>
```
this will wait until the player is alive again. but allows the loop to break if they leave the server
if your intention is to teleport the player to a new respawn location it could jitter (but i never notice it) for 1 tick while they go from their normal respawn position to where you want them to go

**Comment by bloggy** [comment]
It's an idea, but not really a good solution I'd like to use on my server. I hope bensku can finally fix the on respawn event.

**Comment by bloggy** [comment]
The delay after the "on respawn" event is causing lags on a busy server. Why can't this be fixed in Skript?

**Comment by ShaneBeee** [comment]
As stated by Bensku and Snow, its not an issue with Skript, its an issue with Spigot.
In a sense, Skript just mirrors the event from Spigot. So Spigot has to fire the event, then Skript catches it. Im not sure why you think this is a Skript issue. Skript isn't able to fire an event if Spigot hasn't fired it yet, therefor it goes to show that Spigot is firing it too early. 

A note: 
A lot, if not most, of Spigot events, are actually fired before the event takes place. Reason being, so that if you need to cancel said event, it can be done before the event actually occurs.

If you are experiencing lag, its most likely due to inefficient scripts, or issues with your server. 
This is not me trying to defend Skript, this is me reiterating the fact that not everything is Skript's fault, but often a fault of Spigot. 
Some of these issues people have, plugin devs are having the same problems, but the plugin dev finds a work around (ie: waiting a tick or 2), this is just how we deal with odd behaviour on a Spigot server. (Trust me, it happens a LOT)

**Comment by bloggy** [comment]
@ShaneBeee The problem is that the "wait 1 tick" in the "on respawn" event is showing up almost on top in the timings report when lots of users (5 - 30) are respawning at the same time. I spoke with a java dev and he said that he never used something like "wait x ticks" or similar in a on respawn event. So I thought it would be skript's fault. But ofc I don't know for sure. 
Only thing I know is that I have to use "wait 1 tick" in the on respawn event for some effects (e.g. send "test" to event-player etc.) to make them happen.

**Comment by Wealthyturtle** [comment]
It's true that the on respawn event can cause lag, but I don't believe it can be substantial enough to cause severe tps drops. I've seen that other plugins like InstantRespawn also have seemingly high server resource usage (3% at max) for just instant respawning, and I don't think Skript will differ too much from that... Attach your timings report? Something else could be causing the lag.

**Comment by ShaneBeee** [comment]
Im not entirely sure of the convo between you and that dev, but as I explained, Spigot fires the event, which then tells Skript to fire the event. Theres no possible way Skript would fire it before Spigot, therefor its a Spigot issue.

Also, just to show you it's a Spigot issue, I wrote a quick bit of code in java, to test if the player is alive when the PlayerRespawnEvent is called, results:
![Screen Shot 2019-10-10 at 1 32 51 AM](https://user-images.githubusercontent.com/20008482/66552286-e7e9de80-eafd-11e9-9769-855fa0959358.png)

You will see that when the event is called the player is in fact, not alive (see console)
I added a 1 tick wait, after that tick, the player is in fact alive.
